src/stateStream.spec.ts

```diff
@@ -0,0 +1,134 @@
+import { reducer, actionCreator } from 'rxbeach';
+import untypedTest from 'ava';
+import { marbles } from 'rxjs-marbles/ava';
+import sinon from 'sinon';
+import { stateStream, stateStreamWithErrorHandling } from 'rxbeach/stateStream';
+import { of } from 'rxjs';
+import { stubRethrowErrorGlobally } from 'rxbeach/internal/testUtils';
+
+const incrementOne = actionCreator('[increment] one');
+const decrement = actionCreator('[increment] decrement');
+const incrementMany = actionCreator<number>('[increment] many');
+
+const incrementOneHandler = sinon.spy((accumulator: number) => accumulator + 1);
+const handleOne = reducer(incrementOne, incrementOneHandler);
+const handleMany = reducer(
+  incrementMany,
+  (accumulator: number, increment) => accumulator + increment
+);
+const test = stubRethrowErrorGlobally(untypedTest);
+
+const actions = {
+  '1': incrementOne(),
+  '2': incrementMany(2),
+  d: decrement(),
+};
+const words = {
+  a: '1',
+  b: '12',
+};
+const numbers = {
+  '1': 1,
+  '2': 2,
+  '3': 3,
+  '4': 4,
+  '5': 5,
+  '6': 6,
+  '7': 7,
+};
+
+test(
+  'stateStream subscribes action$',
+  marbles(m => {
+    const defaultState = 3;
+    const action$ = m.hot('', actions);
+    const state$ = stateStream('testStream', action$, defaultState, []);
+    state$.subscribe();
+    m.expect(action$).toHaveSubscriptions(['^']);
+  })
+);
+
+test(
+  'stateStream should emit initial state synchronously upon subscription',
+  marbles(m => {
+    const defaultState = 3;
+    const action$ = m.hot('');
+    const expected$ = m.hot('3', numbers);
+    const state$ = stateStream('testStream', action$, defaultState, []);
+
+    m.expect(state$).toBeObservable(expected$);
+  })
+);
+
+test(
+  'stateStream should call reducers synchronously if they emit at once',
+  marbles(m => {
+    const action$ = m.hot('', actions);
+    const word$ = m.hot('    a', words);
+    const expected$ = m.hot('(12)', numbers);
+    const defaultState = 1;
+
+    const handleWord = reducer(
+      word$,
+      (state: number, word) => state + word.length
+    );
+    const state$ = stateStream('testStream', action$, defaultState, [
+      handleWord,
+    ]);
+
+    m.expect(state$).toBeObservable(expected$);
+  })
+);
+
+test('stateStream subscribers should only see one emission if reducers emit synchronously', async t => {
```

Comment:
This is a "good" side-effect, but per @chriskr 's suggestion, we should understand exactly why it happens 😅

Reply:
I think this test is misguided. `toPromise` resolves once the stream completes. In this test `state$` will complete when `action$` completes.
I would guess `state$` actually emits the `defaultState`, but it's not caught because of the behavior of `toPromise`.

Anyways, let's not use `toPromise` in the tests :)

Reply:
True, it's not just misguided, it's completely at odds with the earlier test `reduceState should call reducers synchronously if they emit at once` 🤦‍♂ 

Removed now